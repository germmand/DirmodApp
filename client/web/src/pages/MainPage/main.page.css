.quotations-panel {
    width: 100%;
}

/*
@media only screen and (max-width: 600px) {
    .quotations-panel {
        width: 100%;
    }
}
*/

.container {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    flex: 1;
}

.spinner-container {
    margin-top: 40px;
    width: 100%;
}

.spinner {
    border: 3px solid #f3f3f3;
    border-radius: 50%;
    border-top: 5px solid #3498db;
    width: 120px;
    height: 120px;
    -webkit-animation: spin 1s ease-in-out infinite; /* Safari */
    animation: spin 1s ease-in-out infinite;
    padding: 50px;
    position: absolute;
    left: calc(50% - 48px);
}

.spinner-text {
    text-align: center;
    position: absolute;
    font-size: 65px;
    left: calc(50% - 9px);
    top: calc(50% + 205px);
}

@-webkit-keyframes spin {
  0% { -webkit-transform: rotate(0deg); }
  100% { -webkit-transform: rotate(360deg); }
}

@keyframes spin {
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
}
