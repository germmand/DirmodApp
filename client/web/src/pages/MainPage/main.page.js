import React from 'react';

import './main.page.css';
import QuotationService from '../../services/quotation.service';

import QuotationPanel from '../../components/QuotationPanel';

class MainPage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      currencies: [{
        code: "BRL",
        quotation: 0.00,
        symbol: "$"
      }, {
        code: "USD",
        quotation: 0.00,
        symbol: "$"
      }, {
        code: "EUR",
        quotation: 0.00,
        symbol: "$"
      }],
      isRefreshing: true,
      refreshingCounter: 5,
    };
    this._isMounted = false;
  }

  async componentDidMount() {
    this._isMounted = true;
    try {
      const [usdRes, eurRes, brlRes] = await Promise.all([
        QuotationService.getQuotation('dolar'),
        QuotationService.getQuotation('euro'),
        QuotationService.getQuotation('real')
      ]);
      this.onUpdateCurrencies([usdRes, eurRes, brlRes]);
    } catch(e) {
      // This try-catch block is placed here in order to
      // prevent UnhandledPromiseRejection exception when unit-testing.
      // -----
      // Here should also be handled the API-related errors.
    }
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  onUpdateCurrencies(currenciesUpdated) {
    const { currencies } = this.state;
    const updatedCurrenciesForState = currencies.map(c => {
      const currency = currenciesUpdated.find(uc => uc.currency === c.code);
      return {
        ...c,
        // Round to two decimals if necessary.
        quotation: Number(`${Math.round(`${currency.price  }e2`)  }e-2`),
      };
    });
    if(this._isMounted) {
      this.setState(state => {
        return {
          ...state,
          currencies: updatedCurrenciesForState,
          isRefreshing: false,
        };
      });
    }
  }

  render() {
    const { currencies, isRefreshing, refreshingCounter } = this.state;

    return (
      <div className="container">
        <div className="quotations-panel">
          {
            currencies.map(c => (
              <QuotationPanel
                key={c.code}
                code={c.code}
                symbol={c.symbol}
                isRefreshing={isRefreshing}
                quotation={c.quotation}
              />
            ))
          }
        </div>
        <div className="spinner-container">
          <div className="spinner">
          </div>
          <span className="spinner-text">{ refreshingCounter }</span>
        </div>
      </div>
    );
  }
}

export default MainPage;