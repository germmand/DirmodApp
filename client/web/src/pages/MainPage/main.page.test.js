import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import QuotationService from '../../services/quotation.service';

import MainPage from './main.page';

jest.mock('../../services/quotation.service');
jest.useFakeTimers();

describe('MainPage Component', () => {
  beforeEach(() => {
    QuotationService.getQuotation.mockClear();
  });

  it('should match snapshot', () => {
    const props = {
    };
    const component = renderer.create(
      <MainPage {...props} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  // NOTE:
  // We want the disableLifecycleMethods set to true so that it doesn't call
  // neither componentDidMount nor componentDidUpdate automatically thus we can alter the state
  // accordingly before doing the actual test, since as of Enzyme V3
  // the shallow API does call the React lifecycle methods.
  // As pretty much all of the operations in MainPage start at mounting (i.e. componentDidMount).
  // --------------

  it('substracts one to refreshingCounter if greater than zero and no request made', () => {
    const wrapper = shallow(<MainPage />, { disableLifecycleMethods: true });
    const refreshingCounter = 5;
    const expectedRefreshingCounter = refreshingCounter - 1;
    wrapper.setState({
      refreshingCounter,
      httpRequestMade: false,
    });
    wrapper.instance().componentDidMount();
    jest.advanceTimersByTime(1000);
    expect(wrapper.state().refreshingCounter).toEqual(expectedRefreshingCounter);
  });

  it('does not call setState if there\'s any http request pending', () => {
    const wrapper = shallow(<MainPage />, { disableLifecycleMethods: true });
    // We check if refreshingCounter retained its previous value because
    // this is the property that's intended to change periodically. If it's zero
    // it'll be updated to 5, otherwise it'll decrease by one.
    // So if httpRequestMade is true, no changes to the state should've been triggered
    // hence refreshingCounter should've retained its value.
    const refreshingCounter = 4;
    wrapper.setState({
      httpRequestMade: true,
      refreshingCounter,
    });
    wrapper.instance().componentDidMount();
    jest.advanceTimersByTime(1000);
    expect(wrapper.state().refreshingCounter).toEqual(refreshingCounter);
  });

  it('calls getQuotation using urlCode', () => {
    const dummyCurrencies = [{
        code: "USD",
        urlCode: "dolar",
        quotation: 23.15,
        symbol: "$",
      }, {
        code: "EUR",
        urlCode: "euro",
        quotation: 54.25,
        symbol: "$",
    }];
    const mockedResponses = dummyCurrencies.map(c => ({
      currency: c.code,
      price: c.quotation + 5,
    }));
    QuotationService.getQuotation.mockImplementation((currency) => {
      const currencyFromState = dummyCurrencies
        .find(c => c.urlCode === currency);
      const mockedResponse = mockedResponses
        .find(m => m.currency === currencyFromState.code);
      return Promise.resolve(mockedResponse);
    });
    const wrapper = shallow(<MainPage />, { disableLifecycleMethods: true });
    wrapper.setState({
      httpRequestMade: false,
      refreshingCounter: 0,
      currencies: dummyCurrencies,
    });
    wrapper.instance().componentDidMount();
    jest.advanceTimersByTime(1000);
    expect(QuotationService.getQuotation).toHaveBeenCalledTimes(dummyCurrencies.length);
    dummyCurrencies.forEach(c => {
      expect(QuotationService.getQuotation).toHaveBeenCalledWith(c.urlCode);
    });
  });

  it('updates quotation into state properly', (done) => {
    const dummyCurrencies = [{
      code: "USD",
      urlCode: "dolar",
      quotation: 23.15,
      symbol: "$",
    }, {
      code: "EUR",
      urlCode: "euro",
      quotation: 54.25,
      symbol: "$",
    }];
    const mockedResponses = dummyCurrencies.map(c => ({
      currency: c.code,
      price: c.quotation + 5,
    }));
    const expectedCurrenciesState = dummyCurrencies.map(c => {
      const mockedResponse = mockedResponses
        .find(m => m.currency === c.code);
      return {
        ...c,
        quotation: mockedResponse.price,
      };
    });
    QuotationService.getQuotation.mockImplementation((currency) => {
      const currencyFromState = dummyCurrencies
        .find(c => c.urlCode === currency);
      const mockedResponse = mockedResponses
        .find(m => m.currency === currencyFromState.code);
      return Promise.resolve(mockedResponse);
    });
    const wrapper = shallow(<MainPage />, { disableLifecycleMethods: true });
    wrapper.setState({
      httpRequestMade: false,
      refreshingCounter: 0,
      currencies: dummyCurrencies,
    });
    wrapper.instance().componentDidMount();
    jest.advanceTimersByTime(1000);
    // Since the second state change happens in the setState's callback asynchronously,
    // the operation is placed back onto the event loop and executed on the next tick.
    // By wrapping the assertion in this function, we can guarantee it'll run
    // on the same tick as the second state update and since the event loop follows a FIFO model,
    // the assertion will run *after* the second asynchronous update.
    process.nextTick(() => {
      expect(wrapper.state().currencies).toEqual(expectedCurrenciesState);
      done();
    });
  });
});
