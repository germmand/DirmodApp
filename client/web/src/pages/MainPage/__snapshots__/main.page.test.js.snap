// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage Component should match snapshot 1`] = `
<div
  className="quotations-panel"
>
  <div
    className="card"
  >
    <div
      className="card-body"
    >
      <div
        className="panel"
      >
        <div
          className="flag-code-container"
        >
          <img
            alt="BRL"
            src="BR.svg"
            style={
              Object {
                "display": "inline-block",
                "height": "55px",
                "width": "55px",
              }
            }
          />
          <span
            className="text-bold"
          >
             
            BRL
          </span>
        </div>
        <div
          className="spinner-border"
        />
      </div>
    </div>
  </div>
  <div
    className="card"
  >
    <div
      className="card-body"
    >
      <div
        className="panel"
      >
        <div
          className="flag-code-container"
        >
          <img
            alt="USD"
            src="US.svg"
            style={
              Object {
                "display": "inline-block",
                "height": "55px",
                "width": "55px",
              }
            }
          />
          <span
            className="text-bold"
          >
             
            USD
          </span>
        </div>
        <div
          className="spinner-border"
        />
      </div>
    </div>
  </div>
  <div
    className="card"
  >
    <div
      className="card-body"
    >
      <div
        className="panel"
      >
        <div
          className="flag-code-container"
        >
          <img
            alt="EUR"
            src="EU.svg"
            style={
              Object {
                "display": "inline-block",
                "height": "55px",
                "width": "55px",
              }
            }
          />
          <span
            className="text-bold"
          >
             
            EUR
          </span>
        </div>
        <div
          className="spinner-border"
        />
      </div>
    </div>
  </div>
</div>
`;
